### PR TITLE
feat: add built-in metrics

### PR DESCRIPTION
**Pull request by proost**
previous discussion: https://github.com/redis/rueidis/discussions/199

Add built-in metrics for connection.

I have to otel version up because 1.19.0 doesn't support bucket boundaries.

### CODE REVIEW

File: rueidisotel/metrics.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/metric"
+
+	"github.com/redis/rueidis"
+)
+
+const metricNamespace = "github.com/redis/rueidis"
+
+var DefaultDialLatencyHistogramDefaultBucketBoundaries = []float64{
```
Comment (rueian) [inline rueidisotel/metrics.go:19]:
```suggestion
var DefaultDialLatencyHistogramBuckets = []float64{
```

File: rueidisotel/metrics.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10,
+}
+
+var ErrDialFnRequired = errors.New("rueidisotel: clientOption.DialFn is required")
```
Comment (rueian) [inline rueidisotel/metrics.go:23]: Could we set a default DialFn instead of throwing this error?

File: rueidisotel/metrics.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+type DialLatencyHistogramOption struct {
+	ExplicitBucketBoundaries []float64
```
Comment (rueian) [inline rueidisotel/metrics.go:26]:
```suggestion
	Buckets []float64
```

File: rueidisotel/metrics.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+	}
+}
+
+// Client is a rueidis.Client that tracks connection metrics.
+// It tracks the following metrics:
+// - rueidis_dial_attempt: number of dial attempts
+// - rueidis_dial_success: number of successful dials
+// - rueidis_dial_conns: number of active connections
+// - rueidis_dial_latency: dial latency in seconds
+type Client struct {
```
Comment (rueian) [inline rueidisotel/metrics.go:66]: Well, I think we should reuse and extend the `otelclient` struct.

File: rueidisotel/metrics.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+			Conn:  conn,
+			conns: conns,
+		}, nil
+	}
+}
+
+type connTracker struct {
+	net.Conn
+	conns metric.Int64UpDownCounter
+	once  sync.Once
```
Comment (rueian) [inline rueidisotel/metrics.go:162]:
```suggestion
	once  int32
```

`sync.Once` is overkill. An int32 is enough for this use case.

File: rueidisotel/metrics.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+func (t *connTracker) Close() error {
+	err := t.Conn.Close()
```
Comment (rueian) [inline rueidisotel/metrics.go:166]:
```suggestion
func (t *connTracker) Close() error {
```

File: rueidisotel/metrics.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	t.once.Do(func() {
+		t.conns.Add(context.Background(), -1)
+	})
+
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (rueian) [inline rueidisotel/metrics.go:174]:
```suggestion
	return t.Conn.Close()
```

File: rueidisotel/metrics.go (hunk 1 above)
Comment (proost) [reply rueidisotel/metrics.go]: [aa4b585](https://github.com/redis/rueidis/pull/421/commits/aa4b5858a4916b50a57112948e3ddecc8402df0d)

File: rueidisotel/metrics.go (hunk 2 above)
Comment (proost) [reply rueidisotel/metrics.go]: OK. I Change it. [e026def](https://github.com/redis/rueidis/pull/421/commits/e026def866af3c98a2fbdcc4fdc691b27e0d8da7)

File: rueidisotel/metrics.go (hunk 5 above)
Comment (proost) [reply rueidisotel/metrics.go]: [4171be7](https://github.com/redis/rueidis/pull/421/commits/4171be7c5aaf5b02d52ee8a44c2222b4b58bf25a)

File: rueidisotel/metrics.go (hunk 7 above)
Comment (proost) [reply rueidisotel/metrics.go]: [081c069](https://github.com/redis/rueidis/pull/421/commits/081c06998c5e6eed4f7a01c15810619ee0acf1a2)

File: rueidisotel/metrics.go (hunk 6 above)
Comment (proost) [reply rueidisotel/metrics.go]: [081c069](https://github.com/redis/rueidis/pull/421/commits/081c06998c5e6eed4f7a01c15810619ee0acf1a2)

File: rueidisotel/metrics.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+type connTracker struct {
+	net.Conn
+	conns metric.Int64UpDownCounter
+	once  int32
+}
+
+func (t *connTracker) Close() error {
+	v := atomic.AddInt32(&t.once, 1)
+	if v == 1 {
```
Comment (rueian) [inline rueidisotel/metrics.go:158]:
```suggestion
	if atomic.CompareAndSwapInt32(&t.once, 0, 1) {
```

Even if it is hard to overflow the `t.once`, using `CompareAndSwapInt32` is still better.

File: rueidisotel/metrics.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+var (
+	DefaultDialLatencyHistogramDefaultBuckets = []float64{
+		.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10,
+	}
+	DefaultDialFn = func(dst string, dialer *net.Dialer, _ *tls.Config) (conn net.Conn, err error) {
+		return dialer.Dial("tcp", dst)
+	}
+)
+
+type DialLatencyHistogramOption struct {
```
Comment (rueian) [inline rueidisotel/metrics.go:25]:
```suggestion
type HistogramOption struct {
```
I believe we can reuse this option struct for other histogram metrics. So, name it just `HistogramOption` may be better.

File: rueidisotel/metrics.go [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+	"github.com/redis/rueidis"
+)
+
+var (
+	DefaultDialLatencyHistogramDefaultBuckets = []float64{
+		.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10,
+	}
+	DefaultDialFn = func(dst string, dialer *net.Dialer, _ *tls.Config) (conn net.Conn, err error) {
+		return dialer.Dial("tcp", dst)
+	}
```
[the remark this answers is not in the document]
Comment (proost) [reply rueidisotel/metrics.go]: [dde0cf5](https://github.com/redis/rueidis/pull/421/commits/dde0cf5c1aba63e604a36fd82773872d73fcb889)

File: rueidisotel/metrics.go (hunk 8 above)
Comment (proost) [reply rueidisotel/metrics.go]: [a2b467f](https://github.com/redis/rueidis/pull/421/commits/a2b467ff9c8a0de0dc0dbfca1e2d3646659f6c1c)
Oh, yes. definitely better.

File: rueidisotel/metrics.go (hunk 9 above)
Comment (proost) [reply rueidisotel/metrics.go]: [3895581](https://github.com/redis/rueidis/pull/421/commits/3895581d888aebcbc00fbff76a97e71add2c317d)

File: rueidisotel/metrics.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+// NewClient creates a new Client.
+// The following metrics are recorded:
+// - rueidis_dial_attempt: number of dial attempts
+// - rueidis_dial_success: number of successful dials
+// - rueidis_dial_conns: number of active connections
+// - rueidis_dial_latency: dial latency in seconds
+func NewClient(opts ...Option) (rueidis.Client, error) {
```
[the remark this answers is not in the document]
Comment (proost) [reply rueidisotel/metrics.go]: [ae29088](https://github.com/redis/rueidis/pull/421/commits/ae29088c0a89ef1d1816e75a44bc649824aed020)

Yes. `rueidis.ClientOption` must be required.

**Comment by rueian** [comment]
Looks good to me. Thank you for your contribution! @proost

**Pull request merged**